Objects/dictobject.c

```diff
@@ -242,41 +243,40 @@ static PyObject* dict_iter(PyObject *dict);
 #include "clinic/dictobject.c.h"
 
 
-#if PyDict_MAXFREELIST > 0
+#ifdef WITH_FREELISTS
 static struct _Py_dict_state *
-get_dict_state(PyInterpreterState *interp)
+get_dict_state(void)
 {
-    return &interp->dict_state;
+    _PyFreeListState *state = _PyFreeListState_GET();
+    return &state->dict_state;
 }
 #endif
 
 
 void
-_PyDict_ClearFreeList(PyInterpreterState *interp)
+_PyDict_ClearFreeList(_PyFreeListState *freelist_state, int is_finalization)
 {
-#if PyDict_MAXFREELIST > 0
-    struct _Py_dict_state *state = &interp->dict_state;
-    while (state->numfree) {
+#ifdef WITH_FREELISTS
+    struct _Py_dict_state *state = &freelist_state->dict_state;
+    while (state->numfree > 0) {
         PyDictObject *op = state->free_list[--state->numfree];
         assert(PyDict_CheckExact(op));
         PyObject_GC_Del(op);
     }
-    while (state->keys_numfree) {
+    while (state->keys_numfree > 0) {
         PyObject_Free(state->keys_free_list[--state->keys_numfree]);
     }
+    if (is_finalization) {
+        state->numfree = -1;
+        state->keys_numfree = -1;
+    }
 #endif
 }
 
-
 void
-_PyDict_Fini(PyInterpreterState *interp)
+_PyDict_Fini(_PyFreeListState *freelist_state)
 {
-    _PyDict_ClearFreeList(interp);
-#if defined(Py_DEBUG) && PyDict_MAXFREELIST > 0
-    struct _Py_dict_state *state = &interp->dict_state;
-    state->numfree = -1;
-    state->keys_numfree = -1;
-#endif
+    _PyDict_ClearFreeList(freelist_state, 1);
 }
```

Comment:
Passing anything but `PyInterpreterState *` to `_PyDict_Fini()` is inconsistent and arguably wrong.  Finalization is happening relative to the interpreter, not just some part of it.  It doesn't matter that during finalization `_PyDict_Fini()` only clears its freelist.

There's another problem here.  As it stands, with this PR for no-gil we're clearing the freelist for the current thread rather than all remaining threads.  That can be problematic.  We should do something similar to `_PyGC_ClearAllFreeLists()` but just for dict.

Putting all that together:

```suggestion
_PyDict_Fini(PyInterpreterState *interp)
{
 #ifdef Py_GIL_DISABLED
    HEAD_LOCK(&_PyRuntime);
    _PyThreadStateImpl *tstate = (_PyThreadStateImpl *)interp->threads.head;
    while (tstate != NULL) {
        _PyDict_ClearFreeList(&tstate->freelist_state, 1);
        tstate = (_PyThreadStateImpl *)tstate->base.next;
    }
    HEAD_UNLOCK(&_PyRuntime);
#else
    _PyDict_ClearFreeList(interp->freelist_state);
#endif
}
```

We could simplify that by factoring out helpers that the other freelists have in common here, but that's outside the scope of this PR.

Reply:
@ericsnowcurrently cc @colesbury 
The current implementation is somewhat intended.
Each of thread will be clear own freelist when the thread resources are cleared.
And at the end of interpreter clear phase, interpreter will clear by traversing all threads.
So, I think that there is no issue with the current implementation.

This is somewhat changing with the lifecycle from the free-threading version.

https://github.com/python/cpython/blob/b69548a0f52418b8a2cf7c7a885fdd7d3bfb1b0b/Python/gc_free_threading.c#L1710-L1721

https://github.com/python/cpython/blob/b69548a0f52418b8a2cf7c7a885fdd7d3bfb1b0b/Python/gc_gil.c#L11-L17

Reply:
@ericsnowcurrently 
https://github.com/python/cpython/pull/113584

The relevant discussion was here. I will link the specific comment about this.

Reply:
https://github.com/python/cpython/pull/113584#pullrequestreview-1800574559

Reply:
I agree that the signature is inconsistent, but I don't think we should be clearing freelists for the free-threaded build in `PyDict_Fini()`. The pattern is:

- In the free-threaded build, freelists are per-PyThreadState and cleared in `PyThreadState_Clear()`
- In the default build, freelists are per-interpreter and cleared in `finalize_interp_types()`

By the time we get to `PyDict_Fini()` from `finalize_interp_types()` there are no other thread states in the interpreter and, in the free-threaded build, the freelists for the current thread have already been permanently cleared.

I was going to suggest the following cleanup for after this PR:

1) `finalize_interp_types()` should call `_Py_ClearFreeLists(interp->freelist_state, 1)`. The current pattern has duplicated logic, which makes it prone to bugs, such as a call present in one place and not the other.
2) The call in `finalize_interp_types()` should be guarded by an `#ifndef Py_GIL_DISABLED`.  I think it's better to be explicit that the code there is only useful in the default build.
3) We should remove the functions like `PyDict_Fini()` that only clear freelists and don't have finalize other per-interpreter state. There isn't any per-interpreter finalization left in `PyDict_Fini()`.

Reply:
Thanks for the clarifications!  It would help to have a comment there to indicate why we are only clearing the freelists for the current thread.  As a reader of this code, here are the things it would have been helpful to have known:

* the freelists for any other thread states were already cleared elsewhere (where?)
* (daemon threads aren't a problem because they no longer have valid thread states)
* the mechanism is tied to thread state lifetime rather than thread lifetime (right?)
* the freelists for the current thread state are never cleared by that mechanism (why not?)
* the current thread state when this function is called will always belong to the finalizing interpreter

The comment doesn't need to explain all that.  It just needs enough to keep a reader from asking the same things that led to my review comment.  It could be as simple as:

```suggestion
_PyDict_Fini(_PyFreeListState *freelist_state)
{
    // With Py_GIL_DISABLED:
    // We only need to clear the freelists for the thread state we're using
    // to finalize the current interpreter.  Freelists for all other thread states
    // are cleared elsewhere.
    _PyDict_ClearFreeList(freelist_state, 1);
}
```

It may even make sense to have a comment with a full explanation in `_Py_ClearFreeLists()` and put a short one here pointing there:

```suggestion
_PyDict_Fini(_PyFreeListState *freelist_state)
{
    // With Py_GIL_DISABLED:
    // See `_Py_ClearFreeLists()` about why we only need to clear the freelists
    // for the current thread state.
    _PyDict_ClearFreeList(freelist_state, 1);
}
```

Reply:
> Passing anything but `PyInterpreterState *` to `_PyDict_Fini()` is inconsistent and arguably wrong. Finalization is happening relative to the interpreter, not just some part of it. It doesn't matter that during finalization `_PyDict_Fini()` only clears its freelist.

FYI, I fell the same about this part as I did before.  Let's keep the `void PyDict_Fini(PyInterpreterState *interp)` signature.  (The same goes for the other type-specific `Py*_Fini()` signatures, but we can sort that out in a separate PR.)

Reply:
Yeah, adding comments looks great. People can confuse with the different lifecycles of free lists between the default build and the free threaded build.

I added comments on all related places.
Thanks it's nice :)

Reply:
> the freelists for the current thread state are never cleared by that mechanism

A clarification: in the free-threaded build, the freelists for the current thread state have already been cleared by that mechanism. There are no other thread states [[1]](https://github.com/python/cpython/blob/699779256ec4d4b8afb8211de08ef1382c78c370/Python/pylifecycle.c#L1905) [[2]](https://github.com/python/cpython/blob/699779256ec4d4b8afb8211de08ef1382c78c370/Python/pylifecycle.c#L2238-L2240) (daemon or otherwise). The `_PyDict_ClearFreeList` in `_PyDict_Fini` is redundant for the free-threaded build. You can put an `#ifndef Py_GIL_DISABLED` around it like:

```
_PyDict_Fini(_PyFreeListState *freelist_state)
{
#ifndef Py_GIL_DISABLED
    _PyDict_ClearFreeList(freelist_state, 1);
#endif
}
```

Reply:
> You can put an #ifndef Py_GIL_DISABLED around it like

Done; for other domain, I will handle it with another PR :)

Reply:
Hmm, I applied it in the current PR, since I should add comments for those functions.
https://github.com/python/cpython/pull/114323#discussion_r1467848180